### ISSUE TITLE
Tooltips and popupwindows are clipped by viewport

### ISSUE THREAD

**Issue by MusorGaming** [vurtun/nuklear#144]
Tool-tips and popup windows should be spawned in such a way that they don't get clipped by viewport (when-ever possible).
This is a rather minor UX thing.

**Comment by dumblob** [comment]
I agree with this idea, but I'm afraid target platforms vary quite a lot and therefore a really good, but still lean abstraction will need to be introduced (sounds like callback hell :cry:).

**Comment by MusorGaming** [comment]
dumblob (is this a bot?), What are you talking about?

1. This is not an idea. This is how all the mainstream, polished widget toolkits/UI libraries handle popup windows and tooltips.
2. I can't come up with a single target platform where spawning popupwindows or tooltips so that they intersect with view rectangle would be beneficial.
3. I dont see how this would introduce any sort of new abstractions, other than rather simple logic which would handle the display position of popupwindows and tooltips.

**Comment by dumblob** [comment]
@MusorGaming maybe I misunderstood you - I've been talking about tooltip crossing a real window boundary like this:

![tooltip00](https://cloud.githubusercontent.com/assets/2451044/15517059/0ca91710-21f6-11e6-9b38-dc18d50f45a2.png)

Or have you been talking just about in-canvas tooltip (which is indeed not difficult to handle in Zahnrad without introducing anything new)?

**Comment by vurtun** [comment]
I understood it in context with the OS window which at least for me makes sense.

**Comment by MusorGaming** [comment]
Haha, wow. I didn't even imagine it could be misinterpreted like this.
@dumblob, just the in-canvas tooltips and in-canvas popups of course.
And all that has to be done - is they have to be positioned in a way where they don't clip the *real window*. It's a little bit tricky with popup windows who can have sub-menus spawning to left or right - there you'd have to decide in which direction to spawn the menu. But it's still should be rather simple.

**Comment by dumblob** [comment]
> It's a little bit tricky with popup windows who can have sub-menus spawning to left or right - there you'd have to decide in which direction to spawn the menu. But it's still should be rather simple.

Positioning of tooltips sounds feasible, but with nested menus, it's too much hassle and actually I can name several methods how to handle absence of space (not only deciding whether to spawn them left or right) and therefore I would leave the decision up to the Nuklear user like it's already done with menus (see discussion https://github.com/vurtun/nuklear/issues/119#issuecomment-216272342 ).

**Comment by inolen** [comment]
@vurtun does this happen to be being worked on? It would be extremely useful to have tooltips offset appropriately if they clip with the window.

**Comment by inolen** [comment]
Bumping this again.

@vurtun if a generic fix to this is difficult in the short term, are there any temporary workarounds you can think of? I think possibly just adding a dedicated "debug info preview" group below the list view may be my best bet.

![tracer](https://cloud.githubusercontent.com/assets/460296/21538199/3779e640-cd4f-11e6-8b5f-27f5ea6d82e0.png)
